### PR TITLE
fix(Designer): Updated HTML Editor to support newline and carriage return characters (for dynamic content) 

### PR DESCRIPTION
**Pull request by spanvalkar**
- **Please check if the PR fulfills these requirements**

* [x] The commit message follows our guidelines
* [x] Tests for the changes have been added (for bug fixes/features)
* [x] Docs have been added / updated (for bug fixes / features)

- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

This PR fixes a bug with the HTML Editor where the HTML Editor cannot support newline characters.

- **What is the current behavior?** (You can also link to an open issue here)

When a user adds dynamic content to any action with an HTML Editor and adds a newline character (for example, by pressing the 'enter' key) and then closes the action, reopening the action will cause the dynamic content to disappear. The dynamic content should stay in the action's HTML Editor, but it does not stay because the HTML Editor does not support newline characters.

[Bug 26654763](https://msazure.visualstudio.com/One/_workitems/edit/26654763): Expression gets lost when newline is used

- **What is the new behavior (if this is a feature change)?**

My change enables the HTML Editor to support newline characters. With my change, a user can add newline characters to the HTML Editor for dynamic content and the newline character (and dynamic content) will be preserved. It will not spontaneously disappear (as seen in the bug linked above). This improves the user experience for LA Designer users because it ensures the HTML Editor can support newline characters.

- **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

This PR does not break any existing functionality. Rather, it only adds to the functionality of the HTML Editor by supporting newline characters.

- **Please Include Screenshots or Videos of the intended change**:

Added a newline character after the 'a' in email:

image

Upon reopening the HTML-to-Text action, the dynamic content with newline character remains. Previously, HTML-to-Text field would appear blank (incorrectly):

image

### CODE REVIEW

**Comment by hartra344** [comment]
@Eric-B-Wu  can you take a look at this?

@spanvalkar as discussed offline, I do want to hold off signing off on this until there are adequate tests added. This has been a painful issue in the past.

File: libs/designer-ui/src/lib/editor/base/utils/editorToSegment.ts [hunk 1]
```diff
@@ -37,7 +37,8 @@ const getChildrenNodesToSegments = (node: ElementNode, segments: ValueSegment[],
 export const convertStringToSegments = (
   value: string,
   nodeMap: Map<string, ValueSegment>,
-  options?: SegmentParserOptions
+  options?: SegmentParserOptions,
+  convertSpaceToNewline?: boolean
```
Comment (ek68794998) [inline libs/designer-ui/src/lib/editor/base/utils/editorToSegment.ts:41]: Does it make sense to make this option part of `SegmentParserOptions`?

Comment (spanvalkar) [reply libs/designer-ui/src/lib/editor/base/utils/editorToSegment.ts:41]: Now that you mention it, I think it makes sense to make 'convertSpaceToNewline' a part of the SegmentParserOptions interface. I'll make the change now.

File: libs/designer-ui/src/lib/editor/base/utils/editorToSegment.ts [hunk 2]
```diff
@@ -142,3 +163,19 @@ const collapseLiteralSegments = (segments: ValueSegment[]): void => {
...
   }
 };
+
+const removeNewlinesAndSpaces = (inputStr: string): string => {
+  return inputStr.replace(/\s+/g, '').replaceAll(/\n/g, '').replaceAll(/\r/g, '');
+};
+
+export const getTokenFromNodeMap = (segmentSoFar: string, nodeMap: Map<string, ValueSegment>): string | undefined => {
+  const copyOfSegment = segmentSoFar;
+  const processedId = copyOfSegment.replace(/\n/g, '');
```
Comment (ek68794998) [inline libs/designer-ui/src/lib/editor/base/utils/editorToSegment.ts:173]: For consistency, let's have a method called `removeNewlines` that removes both `\n` and `\r`, and use that on lines 173 and 175 here. `removeNewlinesAndSpaces` can also call that helper.

Comment (spanvalkar) [reply libs/designer-ui/src/lib/editor/base/utils/editorToSegment.ts:173]: Good point. I haven't gone back and optimized my changes yet, so I'll go back and make this change (among a few others I have in mind).

**Comment by Eric-B-Wu** [comment]
asking for tests on the changes made here, as well as a test on a previous regression from this PR [Parameters wiped after saving an update · Issue #4765 · Azure/LogicAppsUX (github.com)](https://github.com/Azure/LogicAppsUX/issues/4765)

**Review by Eric-B-Wu — APPROVED** [review]

**Pull request merged**